### PR TITLE
Improve LegacyTypeDispatch to handle initialization correctly.

### PR DESCRIPTION
**Pull request by ezyang**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #11274 Move the type registry out of Context, into LegacyTypeDispatch.&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D9657449/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#11331 Improve LegacyTypeDispatch to handle initialization correctly.**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D9666612/)

In the previous commit, we added a bare-bones LegacyTypeDispatch in ATen/core.
This is not sufficient for the use cases we need: we not only need to be able to
get a Type, but we also need to be able to *initialize* the Types if its the first time
we have retrieved a CPU/CUDA/Complex type. I hemmed and hawed about how
to do this; the strategy this PR takes is to introduce a new "hooks" interface
specifically for initializing CPU/CUDA/Complex (which still lives in Context). We then
move all "user-friendly" functions to LegacyTypeDispatch.

Here were some other options which I considered, but don't work:
- Assume that Type is already initialized, because we only intend to call Type
  from Tensor methods, where we already have a Tensor. This does not work
  because Caffe2 created tensors will not have gone through the standard
  Type codepath, and will have skipped initialization.
- Move CUDAHooks and ComplexHooks to ATen/core. Besides being sucky,
  this isn't even a complete fix, because I still need to initialize CPU hooks
  (so you *still* need another hooks interface).

Differential Revision: [D9666612](https://our.internmc.facebook.com/intern/diff/D9666612/)

### CODE REVIEW

File: aten/src/ATen/core/LegacyTypeDispatch.cpp [hunk 1]
```diff
@@ -9,4 +9,19 @@ LegacyTypeDispatch & globalLegacyTypeDispatch() {
   return singleton;
 }
 
+AT_DEFINE_REGISTRY(LegacyTypeInitRegistry, LegacyTypeInitInterface, LegacyTypeInitArgs)
+
+const LegacyTypeInitInterface& getLegacyTypeInit() {
+  static std::unique_ptr<LegacyTypeInitInterface> legacy_type_init;
+  static std::once_flag once;
+  std::call_once(once, [] {
```
Comment (cpuhrsch) [inline aten/src/ATen/core/LegacyTypeDispatch.cpp]: why is this necessary?

File: aten/src/ATen/core/LegacyTypeDispatch.h [hunk 2]
```diff
@@ -46,14 +65,83 @@ struct AT_CORE_API LegacyTypeDeleter {
...
+    } else {
+      return getNonVariableType(p, s);
+    }
+  }
   void registerType(Backend b, ScalarType s, TypeUniquePtr&& t) {
     type_registry[static_cast<int>(b)][static_cast<int>(s)] = std::move(t);
     detail::getVariableHooks().registerVariableTypeFor(this, b, s);
   }
 private:
+  void initBackendIfNeeded(DeviceType p) {
```
Comment (cpuhrsch) [inline aten/src/ATen/core/LegacyTypeDispatch.h]: what's the idea behind this?

**Comment by ezyang** [comment]
Caffe2 mobile can't do the initialization, so I concluded that having Caffe2 handle it was not possible (unless you start ifdef'ing.)

File: aten/src/ATen/core/LegacyTypeDispatch.cpp (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/core/LegacyTypeDispatch.cpp]: This is the "standard" boilerplate we've been using for the other hooks mechanisms (e.g., CUDAHooks and VariableHooks). The setting is this: you have been given a registry which, when invoked, constructs an object (whose constructor was specified out of band by some other source). So using this you can construct an object. But you only want to construct the object once, and then reuse it later! Thus `call_once`.

File: aten/src/ATen/core/LegacyTypeDispatch.h (hunk 2 above)
Comment (ezyang) [reply aten/src/ATen/core/LegacyTypeDispatch.h]: LegacyTypeDispatch may not actually have types for CPU or CUDA loaded at the time you try to access them. If they're not loaded (this can only be true the first time we try to initialize them), we have to go request them to actually be loaded. This might involve properly initializing CUDA, for example, which we want to do lazily because it is quite costly.

Comment (gchanan) [inline aten/src/ATen/core/LegacyTypeDispatch.h]: nit: this isn't really a Backend (maybe this is the c10 terminology but it seems strange to use the same name for two different things).  Maybe just avoid this whole issue by calling it "initForDeviceType"?

Comment (ezyang) [reply aten/src/ATen/core/LegacyTypeDispatch.h]: aye aye

**Review by gchanan — APPROVED** [review]
